docs/decisions/0012-system-taxonomy-creation.rst

```diff
@@ -4,25 +4,23 @@
 Context
 --------
 
-System-defined taxonomies are closed taxonomies created by the system. Some of these are totally static (e.g Language)
+System-defined taxonomies are taxonomies created by the system. Some of these are totally static (e.g Language)
 and some depends on a core data model (e.g. Organizations). It is necessary to define how to create and validate 
 the System-defined taxonomies and their tags.
 
 
 Decision
 ---------
 
-System-defined Taxonomy creation
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+System Tag lists and validation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Each System-defined Taxonomy has its own class, which is used for tag validation (e.g. ``LanguageSystemTaxonomy``, ``OrganizationSystemTaxonomy``).
-Each can overwrite ``get_tags``; to configure the valid tags, and ``validate_object_tag``; to check if a list of tags are valid.
-Both functions are implemented on the ``Taxonomy`` base class, but can be overwritten to handle special cases.
+Each System-defined Taxonomy will have its own ``ObjectTag`` subclass which is used for tag validation (e.g. ``LanguageObjectTag``, ``OrganizationObjectTag``).
+Each subclass can overwrite ``get_tags``; to configure the valid tags, and ``is_valid``; to check if a list of tags are valid.  Both functions are implemented on the ``ObjectTag`` base class, but can be overwritten to handle special cases.
```

Comment:
I think this is a conflict in requirements. We need to decide if `ObjectTag` is subclassed/specialized to handle different types of objects (`BlockObjectTag`/`CourseObjectTag`) or different types of Taxonomies (`LanguageObjectTag`/`OrganizationObjectTag`). I don't think it's possible or at least not very clean to have both dimensions of subclass.

Reply:
Personally, I don't think it makes sense to have specialized `ObjectTag` subclasses for System Taxonomies.

I think that special logic belongs in the Taxonomy subclasses, which don't necessarily have to be django model subclasses but can just be proxies or functions.


When displaying tags in the LMS or Studio, we would often want to filter out invalid tags (e.g. from an imported course) or at least display them differently, so I think it's important that `get_object_tags(..., valid_only=True)` is extremely efficient, ideally only using a single database query. For example, if there were a bunch of Author Tags, we wouldn't want each AuthorObjectTag tag making a separate DB query to check if the user still exists in the system. By having a single ObjectTag implementation with the simple logic that you already have in this PR (in `ObjectTag. _check_tag`), we avoid that problem.

One thing to think about is that for the system taxonomies like Language, Author, etc., we don't necessarily have to have them all in the database nor keep them strictly free-form. For Language or Author in particular, the system creates those tags automatically and it could just create the `Tag` for the required language or author if it doesn't exist, right before it creates an `ObjectTag`. Also for Language and Author, we need to do some translating before displaying the tags to the users - displaying the localized name of the language and displaying the username instead of the user ID. I think that sort of translation/lookup should happen in the Taxonomy level or higher so that it can be properly batched, not done on a per-tag basis in case there's several tags of the same type.

Likewise if we ever allow users to manually tag language or author, even if all the possible authors don't exist as Tag instances or free-form values, we could still override an `autocomplete` or `list_available_tags` API for that taxonomy to provide all the possible authors, and then when the ObjectTag is about to be saved, we could create the corresponding `Tag` in the database. But again that feels like behavior that is specific to the taxonomy, not the ObjecTag to me.

Reply:
I like most of the cleanups in this PR, but my recommendation after thinking about it (which you can ignore or push back on!) would be:
* Remove `object_tag_class` and `cast_object_tag`
* Remove references to specialized ObjectTag subclasses
* Consolidate the migrations
And then just merge it as a nice "cleanup" PR.